### DISCUSSION TITLE
Error in Moose Installation

### DISCUSSION THREAD

**Discussion by npaloskey** [idaholab/moose#20672] [Q&A Getting Started] [answered]
I keep getting the following errors when attempting to run tests after compiling libmesh

```
misc/check_error.steady_no_converge: Working Directory: /storage/home/n/ngp11/projects/moose/test/tests/misc/check_error
misc/check_error.steady_no_converge: Running command: /storage/home/n/ngp11/projects/moose/test/moose_test-opt -i steady_no_converge.i --error --error-unused --error-override --no-gdb-backtrace
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: Framework Information:
misc/check_error.steady_no_converge: MOOSE Version:           git commit 16505c7 on 2022-03-28
misc/check_error.steady_no_converge: LibMesh Version:         7315cdb99555e4d774fc149b8c659f1e56955053
misc/check_error.steady_no_converge: PETSc Version:           3.16.5
misc/check_error.steady_no_converge: SLEPc Version:           3.16.2
misc/check_error.steady_no_converge: Current Time:            Wed Mar 30 11:56:42 2022
misc/check_error.steady_no_converge: Executable Timestamp:    Wed Mar 30 03:40:32 2022
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: Parallelism:
misc/check_error.steady_no_converge:   Num Processors:          1
misc/check_error.steady_no_converge:   Num Threads:             1
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: Mesh: 
misc/check_error.steady_no_converge:   Parallel Type:           replicated
misc/check_error.steady_no_converge:   Mesh Dimension:          2
misc/check_error.steady_no_converge:   Spatial Dimension:       2
misc/check_error.steady_no_converge:   Nodes:                   121
misc/check_error.steady_no_converge:   Elems:                   100
misc/check_error.steady_no_converge:   Num Subdomains:          1
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: Nonlinear System:
misc/check_error.steady_no_converge:   Num DOFs:                121
misc/check_error.steady_no_converge:   Num Local DOFs:          121
misc/check_error.steady_no_converge:   Variables:               "u" 
misc/check_error.steady_no_converge:   Finite Element Types:    "LAGRANGE" 
misc/check_error.steady_no_converge:   Approximation Orders:    "FIRST" 
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: Execution Information:
misc/check_error.steady_no_converge:   Executioner:             Steady
misc/check_error.steady_no_converge:   Solver Mode:             Preconditioned JFNK
misc/check_error.steady_no_converge:   PETSc Preconditioner:    hypre boomeramg 
misc/check_error.steady_no_converge: 
misc/check_error.steady_no_converge: HDF5-DIAG: Error detected in HDF5 (1.12.1) thread 0:
misc/check_error.steady_no_converge:   #000: H5T.c line 1929 in H5Tcopy(): not a datatype or dataset
misc/check_error.steady_no_converge:     major: Invalid arguments to routine
misc/check_error.steady_no_converge:     minor: Inappropriate type
```

Not sure what the issue is here but the tests stop after about 176 passing and 50 failed.

**Comment by GiudGiud** [comment]
Hello

Looks like an hdf5 error.
Can you please paste the output of `ldd <your_executable (probably moose_test-opt)> ` ?

Guillaume

**Reply by npaloskey** [reply]
yes, below is the output:

(tutorial-env) [ngp11@submit-004 moose]$ ldd /storage/home/ngp11/projects/moose/test/moose_test-opt

Due to MODULEPATH changes, the following have been reloaded:
  1) cmake/3.18.4

	linux-vdso.so.1 =>  (0x00007fffb1fdc000)
	libmoose_test-opt.so.0 => /storage/home/ngp11/projects/moose/test/lib/libmoose_test-opt.so.0 (0x00007f560a3c0000)
	libmoose-opt.so.0 => /storage/home/ngp11/projects/moose/framework/libmoose-opt.so.0 (0x00007f5608306000)
	libpcre-opt.so.0 => /storage/home/ngp11/projects/moose/framework/contrib/pcre/libpcre-opt.so.0 (0x00007f56080db000)
	libhit-opt.so.0 => /storage/home/ngp11/projects/moose/framework/contrib/hit/libhit-opt.so.0 (0x00007f5607e78000)
	libmesh_opt.so.0 => /storage/home/ngp11/projects/moose/scripts/../libmesh/installed/lib/libmesh_opt.so.0 (0x00007f56068e8000)
	libnetcdf.so.13 => /storage/home/ngp11/projects/moose/scripts/../libmesh/installed/lib/libnetcdf.so.13 (0x00007f56065db000)
	libsz.so.2 => /lib64/libsz.so.2 (0x00007f56063d8000)
	libtimpi_opt.so.10 => /storage/home/ngp11/projects/moose/scripts/../libmesh/installed/lib/libtimpi_opt.so.10 (0x00007f56061bc000)
	libhdf5_cpp.so.8 => /lib64/libhdf5_cpp.so.8 (0x00007f5605f67000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f5605d51000)
	libslepc.so.3.16 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libslepc.so.3.16 (0x00007f560579d000)
	libpetsc.so.3.16 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libpetsc.so.3.16 (0x00007f5603dea000)
	libHYPRE-2.23.0.so => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libHYPRE-2.23.0.so (0x00007f56037c9000)
	libstrumpack.so => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libstrumpack.so (0x00007f5602d6a000)
	libsuperlu_dist.so.7 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libsuperlu_dist.so.7 (0x00007f5602a6a000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f560284e000)
	libhdf5_hl.so.200 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libhdf5_hl.so.200 (0x00007f560262c000)
	libhdf5.so.200 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libhdf5.so.200 (0x00007f5601fb0000)
	libparmetis.so => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libparmetis.so (0x00007f5601d71000)
	libmetis.so => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libmetis.so (0x00007f5601b0f000)
	libX11.so.6 => /lib64/libX11.so.6 (0x00007f56017d1000)
	libmpifort.so.12 => /opt/aci/sw/mpich/3.2_gcc-8.3.1-uia/lib/libmpifort.so.12 (0x00007f5601598000)
	libmpi.so.12 => /opt/aci/sw/mpich/3.2_gcc-8.3.1-uia/lib/libmpi.so.12 (0x00007f5601125000)
	libgfortran.so.5 => /lib64/libgfortran.so.5 (0x00007f5600cad000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f5600a97000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f560088f000)
	libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007f5600653000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f560034b000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f5600147000)
	libmpicxx.so.12 => /opt/aci/sw/mpich/3.2_gcc-8.3.1-uia/lib/libmpicxx.so.12 (0x00007f55fff26000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f55ffc24000)
	libgomp.so.1 => /lib64/libgomp.so.1 (0x00007f55ff9fe000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f55ff630000)
	libaec.so.0 => /lib64/libaec.so.0 (0x00007f55ff428000)
	libhdf5.so.8 => /lib64/libhdf5.so.8 (0x00007f55fef82000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f560aefe000)
	libxcb.so.1 => /lib64/libxcb.so.1 (0x00007f55fed5a000)
	libXau.so.6 => /lib64/libXau.so.6 (0x00007f55feb56000)
(tutorial-env) [ngp11@submit-004 moose]$

Thanks for the quick reply Guillaume! Also to note, the installation is for an HPC cluster. Ive installed before and have had no problems,  I had to delete and reinstall due to storage issues and now I'm running into these problems.

**Reply by GiudGiud** [reply]
so
```
libhdf5_hl.so.200 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libhdf5_hl.so.200 (0x00007f560262c000)
libhdf5.so.200 => /storage/home/n/ngp11/projects/moose/petsc/arch-moose/lib/libhdf5.so.200 (0x00007f5601fb0000)
```
and
```
libhdf5.so.8 => /lib64/libhdf5.so.8 (0x00007f55fef82000)
```
means we are linking to two different hdf5 libraries, which is a recipe for trouble. 

You ll have to re-install and tell petsc where the hdf5 library is (set HDF5_DIR I or _ROOT I think), or find a way to hide the systemh df5 library. 
@milljm do you know how to do that?

**Reply by npaloskey** [reply]
do you know which hdf5 library would be the correct one to use?

**Reply by GiudGiud** [reply]
We might have restrictions on versions.
@cticenhour will know too

**Reply by milljm** [reply]
The correct library will be the one provided by PETSc. Preventing the discovery of system hdf5 though... I sorta thought we had that figured out in libMesh. I'm afraid I won't be of much help on the latter...

**Reply by npaloskey** [reply]
Attempting to install again but this time placing Petsc outside of moose in another directory i created to see if it changes anything. It looks like at the bottom of the code pasted below that PETSC did not detect a system hdf5 and is using its own so i think the problem lies in the libmesh compilation if the same error persists once I compile libmesh again. I will provide updates soon

```pre
(tutorial-env) [ngp11@submit-006 moose]$ ./scripts/update_and_rebuild_petsc.sh --prefix=/storage/home/ngp11/projects/petsc_install

Due to MODULEPATH changes, the following have been reloaded:
  1) cmake/3.18.4

/storage/home/ngp11/projects/moose/scripts
Submodule 'petsc' (https://gitlab.com/petsc/petsc.git) registered for path 'petsc'
Cloning into 'petsc'...
remote: Enumerating objects: 1002562, done.
remote: Counting objects: 100% (45790/45790), done.
remote: Compressing objects: 100% (5880/5880), done.
remote: Total 1002562 (delta 43771), reused 40586 (delta 39906), pack-reused 956772
Receiving objects: 100% (1002562/1002562), 294.59 MiB | 45.08 MiB/s, done.
Resolving deltas: 100% (769619/769619), done.
Submodule path 'petsc': checked out 'f855b95493736b087b8ccc16dc6c5b29bc4b5aa8'
INFO: Checking for HDF5...
INFO: HDF5 library not detected, opting to download via PETSc...
=============================================================================================
                      Configuring PETSc to compile on your system 
```

**Reply by GiudGiud** [reply]
That’s a good point. It could be libmesh or even moose (doubt it very much) looking at the wrong hdf5. Let s build the test executables for petsc and libmesh then check with ldd

**Reply by npaloskey** [reply]
sounds good, it takes like 2ish hours to compile libmesh so ill get back to you when thats completed. Thanks again!

**Reply by npaloskey** [reply]
@GiudGiud, looks like the same issue again, all the same tests are failing. Do you know where I could find the test executables for libmesh and petsc so we can check with ldd?

**Reply by npaloskey** [reply]
Found this in the March MOOSE newsletter, could this have something to do with the issue here?

HDF5 support in PETSc and libMesh now requested by default
The update_and_rebuild_petsc script now performs a basic search for a supported HDF5 installation in the local environment by default. Developers can place an HDF5 installation in the following searched locations:

/usr/lib/hdf5

/usr/local/hdf5

/usr/share/hdf5

/usr/local/hdf5/share

$HOME/.local

Or select a custom location by setting one of the environment variables HDF5_DIR, HDF5DIR, or HDF5_ROOT. If a compatible installation cannot be found, PETSc will download and build it for the user on Linux. This fallback download option is not yet available on MacOS, but will be coming soon. Subsequently, the update_and_rebuild_libmesh script now uses the --enable-hdf5 configure flag and, with recent libMesh changes, attempts to find the HDF5 installation used by PETSc.

The new conda packages mentioned above all contain the PETSc default HDF5 on Linux, but no HDF5 is currently provided on MacOS (coming soon). A serial, pre-built HDF5 is available within the moose-tools package on either platform if this is sufficient for developer needs.

**Reply by npaloskey** [reply]
@GiudGiud 
I think Im starting to put the pieces together. My HPC Cluster system has HDF5 installed (not sure where) but PETSC can't find it any of the five searched locations given in the newsletter. So I think you're correct in saying that I either need to set HDF5_DIR or put the installation in $HOME/.local so PETSC can find it. Im not too sure how to find the HDF5 installation on my system though. Any suggestions? Does this sound right?

**Reply by GiudGiud** [reply]
It has to do with the issue in that new that we install petsc with hdf5, the conflict you are getting appears. 
A few options for setting HDF5_DIR:
- ask your cluster admin where they installed HDF5 so you can point to it. 
- Is there an HDF5 module you can load on your HPC system? That will be easy to locate
- Install your own HDF5 then set the environment variable to that directory

**Reply by npaloskey** [reply]
Yes to bullet point 2. I can do:

load module hdf5/1.12.0

Im pretty sure this just links the hdf5 libraries to the PATH environment variable though. I'm not too sure what I would specify HDF5_DIR as in this case.

**Reply by GiudGiud** [reply]
does it set any of the HDF5_ environment variables to do that? Try auto-complete or `env` to check

**Reply by GiudGiud** [reply]
The modules usually have folders. If you load a module with an executable you can use `which <exec>` to find where it is. Then you can locate the hdf5 folder, and set the environment variable to it

**Reply by npaloskey** [reply]
This is what I get from the module show command:

[ngp11@submit-004 ~]$ module show hdf5/1.12.0
----------------------------------------------------------------------------
   /opt/aci/modulefiles/linux-rhel7-x86_64/gcc/8.3.1/hdf5/1.12.0.lua:
----------------------------------------------------------------------------
whatis("Name : hdf5")
whatis("Version : 1.12.0")
whatis("Target : ivybridge")
whatis("Description : HDF5 is a data model, library, and file format for storing
 and managing data. It supports an unlimited variety of datatypes, and is design
ed for flexible and efficient I/O and for high volume and complex data. ")
whatis("Configure options : --enable-unsupported --enable-symbols=yes --disable-
threadsafe --disable-cxx --enable-hl --enable-fortran --disable-java --without-s
zlib --enable-build-mode=production --enable-shared CFLAGS=-fPIC CXXFLAGS=-fPIC 
FCFLAGS=-fPIC --with-zlib=/usr")
help([[HDF5 is a data model, library, and file format for storing and managing
data. It supports an unlimited variety of datatypes, and is designed for
flexible and efficient I/O and for high volume and complex data.]])
prepend_path("LD_LIBRARY_PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/lib")
prepend_path("LIBRARY_PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/lib")
prepend_path("PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/bin")
prepend_path("C_INCLUDE_PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/include")
prepend_path("CPLUS_INCLUDE_PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/include
")
prepend_path("INCLUDE","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/include")
prepend_path("CMAKE_PREFIX_PATH","/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb/")

[ngp11@s

**Reply by GiudGiud** [reply]
There's your HDF5_DIR to set: `/opt/aci/sw/hdf5/1.12.0_gcc-8.3.1-rpb`

I hope 1.12 is new enough

**Reply by npaloskey** [reply]
Great, I hope so too haha. Ill let you know how it goes. Thanks again

**Reply by GiudGiud** [reply]
To check the hdf5 in libmesh btw:
```
cd libmesh/build/.libs
ldd libmesh_opt.so
```

**Reply by npaloskey** [reply]
Hey @GiudGiud, I think what we did worked. 3050 tests passed and only 10 failed when doing 

cd ~/projects/moose/test
make -j 4
./run_tests -j 4

Thanks for the help!!

**Reply by GiudGiud** [reply]
I’m glad! 
if you please could paste the full error messages for these ten tests, could help us find an issue. 
‘./run_tests —failed-tests’ will only run these ten

**Reply by npaloskey** [reply]
Hey @GiudGiud , I think some of the tests call for the system default version of python, my system defaults to python 2.7 for some reason so I created a virtual environment that would default my system to using python 3.6.8 and now all tests are passing. 0 failed. I think we're all good here.

Thanks again!

**Comment by npaloskey** [comment, marked as answer]
Set HDF5_DIR env variable to location of HDF5 installation in system

**Reply by milljm** [reply]
I was unaware of HDF5_DIR as an influential environment variable. Good to know, and I probably should have :) Cheers!

**Comment by jessecarterMOOSE** [comment]
Thanks all. Had this same problem. Fixed by using system HDF5, which was 1.10.7.
